src/main/java/com/gtnewhorizons/modularui/api/math/MathExpression.java

```diff
@@ -1,315 +1,593 @@
 package com.gtnewhorizons.modularui.api.math;
 
-import java.text.ParseException;
+import java.text.NumberFormat;
+import java.text.ParsePosition;
 import java.util.ArrayList;
-import java.util.Collections;
 import java.util.List;
+import java.util.function.BiFunction;
+import java.util.regex.Pattern;
 
-import com.gtnewhorizons.modularui.common.widget.textfield.TextFieldWidget;
+import org.jetbrains.annotations.NotNull;
 
 public class MathExpression {
 
-    private static final List<Object> DEFAULT = Collections.singletonList(0);
+    /**
+     * Matches any string that can be evaluated to a number. The pattern might be too generous, i.e., matches some
+     * strings that do not evaluate to a valid value. See {@link #parseMathExpression(String, Context)} for an
+     * explanation of the syntax.
+     */
+    public static final Pattern EXPRESSION_PATTERN = Pattern.compile("[0-9.,  +\\-*/^()eEkKmMgGbBtT%]*");
+    // Character   to support French locale thousands separator.
+
+    // TODO:
+    // There should be a config option to default the number format to either the user's computer locale, or to a
+    // locale specified in a config. This way, users who have their computer locale set to a different one still have an
+    // option to input numbers using English conventions.
```

Comment:
The ultimate solution would be completely separating the value stored in server/client widget and the text shown on client. That way player can customize formatting however they want, while server still knowing the actual value without formatting. However, it's way beyond the scope of this PR, and I doubt it could be made possible without breaking a lot of things. Maybe it could be easier with MUI2.